sound/soc/sof/ipc.c

```diff
@@ -491,12 +491,49 @@ static void ipc_xrun(struct snd_sof_dev *sdev, u32 msg_id)
 #endif
 }
 
+static int kwd_notification_handler(struct snd_sof_dev *sdev, u32 comp_id)
+{
+	int ret = -ENODEV;
+	int direction = SNDRV_PCM_STREAM_CAPTURE;
+	int cmd = SNDRV_PCM_TRIGGER_START;
+	struct snd_sof_pcm *spcm;
+	struct snd_pcm_substream *substream;
+
+	/* Firmware detected a key phrase and is about to start draining.
+	 * Its timme to resume the KWD stream now.
+	 *
+	 * Find correct stream first - KWD stream has suspend_ignored
+	 * attribute assigned.
+	 */
+	list_for_each_entry(spcm, &sdev->pcm_list, list) {
+		if (spcm->stream[direction].suspend_ignored) {
+			substream = spcm->stream[direction].substream;
+			ret = snd_sof_pcm_platform_trigger(sdev,
+							   substream,
+							   cmd);
```

Comment:
Doesn't this break the assumption that we do not initiate an IPC while handling an IPC? We've done this before and it deadlocked.

Reply:
@plbossart The snd_sof_pcm_platform_trigger() here is only setting registers, no extra IPC will be sent to the FW with it.

Reply:
@keyonjie @mrajwa there must be a notification to the FW with the reply to the IPC saying draining can being now isnt it?

Reply:
@keyonjie @ranj063  ack - we need to reply to the notification with result of the platform_trigger(). FW must wait for platform_trigger() success before it drains.

Reply:
@lgirdwood, but I am not sure if we really need this resume ack IPC from host. The suspend trigger will stop the DMA engine associated with this stream and HOST component on the DSP side checks its state therefore it won't copy further if gateway buffer is full - and it will be until we resume the DMA.

Reply:
@mrajwa may I suggest baby-steps, one at a time. First we try to disable the DMAs in suspend and renable them in resume, then if this works, check on the latency and if this is not good enough try to improve further with an IPC-based implementation. That way we can make multiple drops and have revalidation of changes.

Reply:
And rethinking about this @mrajwa please be careful - you do need a synchronization. It could happen that a keyword is detected during the suspend routine, so you could actually start pushing data out since the DMAs are still enabled, and while that happens the DMAs get disabled as part of the suspend routine. That's pretty scary. 

Regardless of the solution used for resume, on suspend you also need to detect if you receive the PM_GATE IPC and reject it if you've already started streaming so that we don't disable those DMAs if they are in use.

Reply:
> @mrajwa may I suggest baby-steps, one at a time. First we try to disable the DMAs in suspend and renable them in resume, then if this works, check on the latency and if this is not good enough try to improve further with an IPC-based implementation. That way we can make multiple drops and have revalidation of changes.

This approach seems reasonable to me. Since there is a potential race risk on synchronization via IPC mechanism then we should focus on system trigger resume as a first step. Next we should continue validation and hardening of solution proposed by @mrajwa to optimize resume latency.

Reply:
Actually there is also a race on suspend... We need to fix the suspend case regardless of the resume implementation.

problematic sequence:
Driver starts the suspend handler
Firmware detects a sequence and starts the DMA
Driver disables the DMAs.

We really need a handshake where the firmware rejects the DMA stop notification if the DMA is already in-use